suppression-backup-service/cmd/main.go

```diff
@@ -0,0 +1,139 @@
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+	"github.com/rudderlabs/rudder-server/services/diagnostics"
+	"github.com/rudderlabs/rudder-server/services/stats"
+	"github.com/rudderlabs/rudder-server/suppression-backup-service/api"
+	"github.com/rudderlabs/rudder-server/suppression-backup-service/exporter"
+	"github.com/rudderlabs/rudder-server/suppression-backup-service/model"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"golang.org/x/sync/errgroup"
+)
+
+var pkgLogger = logger.NewLogger().Child("suppression-backup-service")
+
+func main() {
+	pkgLogger.Info("Starting suppression backup service")
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	err := Run(ctx)
+	if ctx.Err() == nil {
+		cancel()
+	}
+	if err != nil {
+		pkgLogger.Error(err)
+		os.Exit(1)
+	}
+}
+
+func Run(ctx context.Context) error {
+	err := stats.Default.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("could not start stats: %w", err)
+	}
+	defer stats.Default.Stop()
+
+	id, err := getIdentity(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get identity: %w", err)
+	}
+	exportBaseDir, err := exportPath()
+	if err != nil {
+		return fmt.Errorf("could not get export path: %w", err)
+	}
+
+	fullExportFile := model.File{Path: path.Join(exportBaseDir, "full-export"), Mu: &sync.RWMutex{}}
+	latestExportFile := model.File{Path: path.Join(exportBaseDir, "latest-export"), Mu: &sync.RWMutex{}}
+
+	g, gCtx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		fullExporter, err := exporter.NewExporter(id, fullExportFile, pkgLogger, exporter.WithPollIntervalFn(func() time.Duration {
+			return config.GetDuration("SuppressionExporter.fullExportInterval", 24*60*60, time.Second)
+		}))
+		if err != nil {
+			return fmt.Errorf("could not create full exporter: %w", err)
+		}
+		return fullExporter.FullExporterLoop(gCtx)
+	})
+
+	g.Go(func() error {
+		latestExporter, err := exporter.NewExporter(id, latestExportFile, pkgLogger)
+		if err != nil {
+			return fmt.Errorf("could not create latest exporter: %w", err)
+		}
+		return latestExporter.LatestExporterLoop(gCtx)
+	})
+
+	var httpServer *http.Server
+	g.Go(func() error {
+		api := api.NewAPI(pkgLogger, fullExportFile, latestExportFile)
+		httpServer = &http.Server{
+			Addr:              fmt.Sprintf(":%s", config.GetString("HTTP_PORT", "8000")),
+			Handler:           api.Handler(gCtx),
+			ReadHeaderTimeout: config.GetDuration("HTTP_READ_HEADER_TIMEOUT", 3, time.Second),
+		}
+		return httpServer.ListenAndServe()
+	})
+	pkgLogger.Info("http server setup done")
+
+	g.Go(func() error {
+		<-gCtx.Done()
+		pkgLogger.Info("context.Done triggered... shutting down httpserver")
+		return httpServer.Shutdown(context.Background())
+	})
+
+	if err := g.Wait(); err != nil {
+		if err != http.ErrServerClosed || ctx.Err() != nil {
```

Comment:
I would argue that `if ctx.Err() != nil` we shouldn't return an `error` under any condition, because this is the definition of a normal shutdown.

Reply:
modified it to `err == http.ErrServerClosed && ctx.Err() == nil `, which means that since ctx.Err==nil means `Done()` is not closed yet. And, still server is closed. So, it is unexpected.